Dockerfile

```diff
@@ -15,6 +15,28 @@
 # This is a standard practice in the Docker world.
 # See https://docs.docker.com/build/building/multi-stage/
 
+###############################################################################
+# Step0: collect files needed to build semgrep-cpre
+###############################################################################
+
+# The semgrep git repository contains the source code to multiple build artifacts
+# (semgrep, semgrep-core, semgrep.js, etc...). In order to maximize Docker cache
+# hits (and keep the build fast), we only copy over the folders needed to build
+# semgrep-core. This is done in a multi-stage build so that the final COPY
+# happens in a single layer.
+
+FROM busybox:stable as semgrep-core-files
+WORKDIR /src/semgrep
+
+# copy over the entire semgrep repository
+COPY . .
+
+# remove folders that aren't necessary for the semgrep-core build
```

Comment:
Not a fan. Is it worth it? Now if in the futur we add more directories in semgrep we need to modify this list ... You avoid the docker build time for PRs that modify the cli/ and js/ which are rare, so is it really worth the added complexity?

Reply:
why not adding stats/ to this list? what about perf/? Seems like an arbitrary list that is bound to be obsolete in a few weeks and not really optimizing anything.

Reply:
Or is it more useful because we started to switch to depot.dev and the optimization is more visible?

Reply:
Yep, these changes have been a huge help for me as I've iterated on changes to GHA, python wheels, semgrep.js, etc... The docker build has done down from minutes to ~10 sec if everything is cached.

You're right that there are likely a few more folders that I could exclude, but I wanted to stick with the folders that I knew for certain weren't necessary for the docker build. It is unfortunate that this list will become out-of-date over time, but the way I see it:
- if we start needing a folder that is deleted, the build will fail and we'll fix it
- if a new folder is added that is causing unacceptable slowness, someone will get annoyed and we'll fix it